### PR TITLE
feat: severity setting [ROAD-1143]

### PR DESCRIPTION
**Pull request by j-luong**
### Description

_This PR should implement filtering based on severity._

### Checklist

- [x] Tests added and all succeed
- [x] Linted

### CODE REVIEW

File: application/config/config.go [hunk 1]
```diff
@@ -146,6 +146,10 @@ type Config struct {
 	integrationVersion          string
 	automaticAuthentication     bool
 	tokenChangeChannels         []chan string
+	filterCriticalSeverity      concurrency.AtomicBool
```
Comment (michelkaporin) [inline application/config/config.go:149]: can we have a complex object in settings for this? That'd nicely group some of the settings :)

Comment (j-luong) [reply application/config/config.go]: done

File: application/config/config.go [hunk 2]
```diff
@@ -188,6 +190,7 @@ func New() *Config {
 	c.clientSettingsFromEnv()
 	c.deviceId = c.determineDeviceId()
 	c.addDefaults()
+	c.setDefaultSeverityFilter()
```
Comment (bastiandoetsch) [inline application/config/config.go:193]: why not fold this into `addDefaults()`?

File: application/server/configuration.go [hunk 3]
```diff
@@ -223,3 +224,19 @@ func updateProductEnablement(settings lsp.Settings) {
...
+	emptySeverityFilter := lsp.SeverityFilter{}
+	defaultSeverityFilter := lsp.SeverityFilter{
+		Critical: true,
+		High:     true,
+		Medium:   true,
+		Low:      true,
+	}
+
+	if s == emptySeverityFilter {
+		config.CurrentConfig().SetSeverityFilter(defaultSeverityFilter)
```
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply application/server/configuration.go:238]: Had a look at the acceptance criteria, unfortunately this scenario is missing.

File: domain/ide/workspace/folder.go [hunk 4]
```diff
@@ -134,20 +135,53 @@ func (f *Folder) DocumentDiagnosticsFromCache(file string) []snyk.Issue {
...
+
 		f.documentDiagnosticCache.Put(issue.AffectedFilePath, cachedIssues)
-		issuesByFile[issue.AffectedFilePath] = cachedIssues.([]snyk.Issue)
+	}
+
+	// update issues by file
+	filteredIssues := []snyk.Issue{}
+	severityFilters := config.CurrentConfig().FilterSeverity()
+	// prevent panic: interface conversion: interface {} is nil, not []snyk.Issue
+	if cachedIssues == nil {
```
Comment (bastiandoetsch) [inline domain/ide/workspace/folder.go:161]: So, in which constellation can this be nil? If there is more than one issue to be processed, the array is created in line 147. If `len(issues)` is 0, why are we even processing? Shouldn't we just early end the processing and return

File: domain/ide/workspace/folder_test.go [hunk 5]
```diff
@@ -165,6 +167,57 @@ func TestProcessResults_whenSamePathsAndDuplicateIssues_DeDuplicates(t *testing.
...
+		{ID: "id3", AffectedFilePath: "path1", Severity: snyk.Medium},
+		{ID: "id4", AffectedFilePath: "path1", Severity: snyk.Low},
+		{ID: "id5", AffectedFilePath: "path1", Severity: snyk.Critical},
+	})
+
+	mtx := &sync.Mutex{}
+	var diagnostics []lsp.Diagnostic
+
+	defer notification.DisposeListener()
+	notification.CreateListener(func(event interface{}) {
```
Comment (bastiandoetsch) [inline domain/ide/workspace/folder_test.go:197]: why not use the jsonRPCRecorder?

File: application/server/configuration.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (j-luong) [reply application/server/configuration.go:238]: It's to guard against the second thing; when no severity setting is passed/set. This was the case in some of the tests; but I think in reality, the client should always pass the initial severity settings when the LS client is created.

Actually, we can add the `if s == lsp.SeverityFilter{} { return }` in the `config.SetSeverityFilter()` instead, which would be cleaner.

File: domain/ide/workspace/folder.go (hunk 4 above)
Comment (j-luong) [reply domain/ide/workspace/folder.go:161]: It looks like this was a remnant from earlier in the development, just tried it without this check and it's working fine 👍 
If `len(issues) == 0` then we'd skip the filtering and process the diagnostics/hovers as we did previously.

File: domain/ide/workspace/folder_test.go (hunk 5 above)
Comment (j-luong) [reply domain/ide/workspace/folder_test.go:197]: I'm not familiar enough with it tbh and just followed the example set in the test below this one :)

File: domain/ide/workspace/folder.go [hunk 6]
```diff
@@ -142,20 +142,54 @@ func (f *Folder) DocumentDiagnosticsFromCache(file string) []snyk.Issue {
 }
 
 func (f *Folder) processResults(issues []snyk.Issue) {
-	var issuesByFile = map[string][]snyk.Issue{}
+	if len(issues) == 0 {
+		f.documentDiagnosticCache.ClearAll()
+		return
+	}
```
Comment (asaf92) [inline domain/ide/workspace/folder.go:148]: This is a bug. 
This function can be called from a scan of a sub-folder/specific file, but the cache holds issues for all files/folders. and the results get cleared

Comment (j-luong) [reply domain/ide/workspace/folder.go]: 👍 this will be removed once #192 is merged, which will make this line redundant

**Review by asaf92 — APPROVED** [review]

**Pull request merged**